contracts/BSCValidatorSet.sol

```diff
@@ -470,15 +563,182 @@ contract BSCValidatorSet is IBSCValidatorSet, System, IParamSubscriber, IApplica
+        ));
         currentValidatorSetMap[validatorSet[i].consensusAddress] = i+1;
       }
     }
+
+    // make sure all new validators are cleared maintainInfo
+    // should not happen, still protect
+    for (uint i = 0; i < currentValidatorSet.length; i++) {
+      validatorExtraSet[i].isMaintaining = false;
+      validatorExtraSet[i].enterMaintenanceHeight = 0;
+    }
   }
 
   function isSameValidator(Validator memory v1, Validator memory v2) private pure returns(bool) {
     return v1.consensusAddress == v2.consensusAddress && v1.feeAddress == v2.feeAddress && v1.BBCFeeAddress == v2.BBCFeeAddress && v1.votingPower == v2.votingPower;
   }
 
+  function _misdemeanor(address validator) private returns (uint256) {
+    uint256 index = currentValidatorSetMap[validator];
+    if (index <= 0) {
+      return ~uint256(0);
+    }
+    // the actually index
+    index = index - 1;
+
+    uint256 income = currentValidatorSet[index].incoming;
+    currentValidatorSet[index].incoming = 0;
+    uint256 rest = currentValidatorSet.length - 1;
+    emit validatorMisdemeanor(validator, income);
+    if (rest == 0) {
+      // should not happen, but still protect
+      return index;
+    }
+    uint256 averageDistribute = income / rest;
+    if (averageDistribute != 0) {
+      for (uint i = 0; i < index; i++) {
+        currentValidatorSet[i].incoming = currentValidatorSet[i].incoming + averageDistribute;
+      }
+      uint n = currentValidatorSet.length;
+      for (uint i = index + 1; i < n; i++) {
+        currentValidatorSet[i].incoming = currentValidatorSet[i].incoming + averageDistribute;
+      }
+    }
+    // averageDistribute*rest may less than income, but it is ok, the dust income will go to system reward eventually.
+
+    return index;
+  }
+
+  function _felony(address validator) private {
+    uint256 index = currentValidatorSetMap[validator];
+    if (index <= 0) {
+      return;
+    }
+    // the actually index
+    index = index - 1;
+    uint256 income = currentValidatorSet[index].incoming;
+    uint256 rest = currentValidatorSet.length - 1;
+    if (getValidators().length <= 1) {
+      // will not remove the validator if it is the only one validator.
+      currentValidatorSet[index].incoming = 0;
+      return;
+    }
+    emit validatorFelony(validator, income);
+
+    // remove the validator from currentValidatorSet
+    delete currentValidatorSetMap[validator];
+    // It is ok that the validatorSet is not in order.
+    if (index != currentValidatorSet.length - 1) {
+      currentValidatorSet[index] = currentValidatorSet[currentValidatorSet.length - 1];
+      validatorExtraSet[index] = validatorExtraSet[currentValidatorSet.length - 1];
+
+      currentValidatorSetMap[currentValidatorSet[index].consensusAddress] = index + 1;
+    }
+    currentValidatorSet.pop();
+    validatorExtraSet.pop();
+
```

Comment:
Please decrease `numOfMaintaining` if this node is maintaining.

Reply:
Please set `numOfMaintaining ` to zero when validator change package arrived.

Reply:
fixed